### PR TITLE
Implement Authenticator

### PR DESCRIPTION
**Pull request by fknop**

### CODE REVIEW

File: app/src/main/assets/github_.css [hunk 1]
```diff
@@ -0,0 +1,391 @@
+body {
```
Comment (Astalaseven) [inline app/src/main/assets/github_.css]: why two different `github.css` ?

File: app/src/main/java/gitlin/kothub/accounts/LoginActivity.kt [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            }
+
+            accountManager.addAccountExplicitly(account, null, data)
+            accountManager.setAuthToken(account, getString(R.string.accountType), token)
+
+            val result = Bundle()
+            with(result) {
+                putString(AccountManager.KEY_ACCOUNT_NAME, account.name)
+                putString(AccountManager.KEY_ACCOUNT_TYPE, account.type)
+                putString(AccountManager.KEY_AUTHTOKEN, token)
```
Comment (Astalaseven) [inline app/src/main/java/gitlin/kothub/accounts/LoginActivity.kt:102]: KEY_AUTH_TOKEN ?

**Review by Astalaseven — COMMENTED** [review]
A good practice could be to right click on the `java` folder in the Project view and select `Optimize Imports`. That way it does it for the whole project.

File: app/src/main/assets/github_.css (hunk 1 above)
Comment (fknop) [reply app/src/main/assets/github_.css]: It was for testing purpose. They are not even used yet.

File: app/src/main/java/gitlin/kothub/github/api/dsl/Edges.kt [hunk 3]
```diff
@@ -1,12 +1,26 @@
 package gitlin.kothub.github.api.dsl
 
-import kotlin.reflect.KClass
+import kotlin.reflect.full.primaryConstructor
 
-abstract class Edges(override val level: Int): Element {
+//abstract class Edges(override val level: Int): Element {
```
Comment (Astalaseven) [inline app/src/main/java/gitlin/kothub/github/api/dsl/Edges.kt]: Should clean those before merge

File: app/src/main/java/gitlin/kothub/github/api/dsl/Issue.kt [hunk 4]
```diff
@@ -11,29 +11,29 @@ class Issue(override val level: Int): CommentFields {
...
-
-
-class IssueConnection(override val level: Int) : Connection<Issue>(level) {
-    fun edges (body: IssueEdges.() -> Unit) {
-        val edges = IssueEdges(nextLevel())
-        edges.body()
-        addField(Node("edges", edges.fields))
-    }
-}
+//class IssueEdges(override val level: Int): Edges(level) {
```
Comment (Astalaseven) [inline app/src/main/java/gitlin/kothub/github/api/dsl/Issue.kt]: Remove commented code before merge

File: app/src/main/java/gitlin/kothub/github/api/dsl/User.kt [hunk 5]
```diff
@@ -122,21 +122,21 @@ class User(override val level: Int): ActorFields {
...
-
-class UserEdge(override val level: Int): Edges(level) {
-    fun node (body: User.() -> Unit) {
-        val issue = User(nextLevel())
-        issue.body()
-        addField(Node(fields = issue.fields))
-    }
-}
-
+//
```
Comment (Astalaseven) [inline app/src/main/java/gitlin/kothub/github/api/dsl/User.kt]: Remove commented code

File: app/src/main/java/gitlin/kothub/ui/LifecycleAppCompatActivity.kt [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+open class LifecycleAppCompatActivity: AppCompatActivity(), LifecycleRegistryOwner, AnkoLogger {
+
+    private val registry = LifecycleRegistry(this)
+
+    override fun getLifecycle(): LifecycleRegistry = registry
+
+    override fun onCreateOptionsMenu(menu: Menu): Boolean {
+        IconicsMenuInflatorUtil.inflate(menuInflater, this, R.menu.base_menu, menu)
+
+//        val searchManager = getSystemService(Context.SEARCH_SERVICE) as SearchManager
```
Comment (Astalaseven) [inline app/src/main/java/gitlin/kothub/ui/LifecycleAppCompatActivity.kt]: Remove commented code

File: app/src/main/java/gitlin/kothub/ui/drawer/Drawer.kt [hunk 7]
```diff
@@ -104,16 +111,18 @@ class AppDrawer(private val activity: AppCompatActivity, toolbar: Toolbar): Life
...
-                val hour = SimpleDateFormat("HH:mm").format(date)
+                    val date = java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'").parse(currentRateLimit?.resetAt)
+                    val hour = java.text.SimpleDateFormat("HH:mm").format(date)
 
-                activity.alert("The rate limit is the number of points remaining on your API account. " +
-                               "It resets every hour. If you somehow reach zero, you will not be able to use the application until $hour GMT+00:00") {
-                    okButton {  }
-                }.show()
+                    activity.alert("The rate limit is the number of points remaining on your API account. " +
+                                   "It resets every hour. If you somehow reach zero, you will not be able to use the application until $hour GMT+00:00") {
```
Comment (Astalaseven) [inline app/src/main/java/gitlin/kothub/ui/drawer/Drawer.kt]: Remove `somehow`, we shouldn't juge the user

File: app/src/main/java/gitlin/kothub/ui/repo/RepositoryActivity.kt [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+        })
+
+        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
+        toolbar.setNavigationOnClickListener {
+            finish()
+        }
+
+
+
+//        fab.setImageDrawable(
```
Comment (Astalaseven) [inline app/src/main/java/gitlin/kothub/ui/repo/RepositoryActivity.kt]: Remove commented code

File: app/src/main/java/gitlin/kothub/ui/repo/RepositoryActivity.kt [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+    }
+
+    private fun updateHeaderView (repo: RepositorySummary) {
+
+        nameWithOwner.text = repo.nameWithOwner
+        Picasso.with(ownerPicture.context).load(repo.ownerAvatarUrl).into(ownerPicture)
+
+        if (repo.language != null) {
+            languageName.text = repo.language.name
+//            languageColor.color = Color.parseColor(repo.language.color)
```
Comment (Astalaseven) [inline app/src/main/java/gitlin/kothub/ui/repo/RepositoryActivity.kt]: Remove commented code

File: app/src/main/java/gitlin/kothub/ui/repo/RepositoryActivity.kt [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+
+        val PAGE_COUNT = 4
+
+        override fun getCount(): Int = PAGE_COUNT
+
+        override fun getItem(position: Int): Fragment = RepositoryFragment.newInstance()
+
+        override fun getPageTitle(position: Int): CharSequence? {
+            return when (position) {
+                0 -> "Home"
```
Comment (Astalaseven) [inline app/src/main/java/gitlin/kothub/ui/repo/RepositoryActivity.kt]: Store const string in strings.xml

File: app/src/main/java/gitlin/kothub/ui/settings/SettingsActivity.kt [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+        toolbar.setNavigationOnClickListener {
+            finish()
+        }
+
+
+        createFragment(savedInstanceState, R.id.preferences_placeholder) {
+            SettingsFragment()
+        }
+
+        //drawer = AppDrawer(this, toolbar)
```
Comment (Astalaseven) [inline app/src/main/java/gitlin/kothub/ui/settings/SettingsActivity.kt]: Remove commented code

File: app/src/main/java/gitlin/kothub/utilities/Fragment.kt [hunk 12]
```diff
@@ -4,11 +4,23 @@ import android.os.Bundle
...
+
+inline fun AppCompatActivity.removeFragment (tag: String) {
+
+    val ft = supportFragmentManager.beginTransaction()
+    val fragment = supportFragmentManager.findFragmentByTag(tag)
+    if (fragment != null) {
+        Log.i("Fragment", "not null")
+        ft.remove(fragment).commit()
+    }
+//    ft.commit()
```
Comment (Astalaseven) [inline app/src/main/java/gitlin/kothub/utilities/Fragment.kt]: Remove commented code

File: app/src/main/java/gitlin/kothub/utilities/markdown/HtmlTagHandler.kt [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+import android.util.Log
+import org.xml.sax.XMLReader
+import java.util.*
+import kotlin.reflect.KClass
+
+
+class HtmlList {
+
+    fun append(output: Editable) {
+        output.append('\u2022')
```
Comment (Astalaseven) [inline app/src/main/java/gitlin/kothub/utilities/markdown/HtmlTagHandler.kt:16]: Can't we use the unicode character `•`?

File: app/src/main/res/layout/repository_header.xml [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+<android.support.constraint.ConstraintLayout
+xmlns:android="http://schemas.android.com/apk/res/android"
+xmlns:app="http://schemas.android.com/apk/res-auto"
+xmlns:tools="http://schemas.android.com/tools"
+android:layout_width="match_parent"
+android:layout_height="wrap_content"
+    tools:layout_editor_absoluteY="25dp"
+    tools:layout_editor_absoluteX="0dp">
+
+    <!--<de.hdodenhof.circleimageview.CircleImageView-->
```
Comment (Astalaseven) [inline app/src/main/res/layout/repository_header.xml]: Remove unused code before merge

File: app/src/main/res/values/styles.xml [hunk 15]
```diff
@@ -1,20 +1,24 @@
...
-        <item name="android:colorPrimary">@color/colorPrimary</item>
-        <item name="android:colorPrimaryDark">@color/colorPrimaryDark</item>
-
-        <item name="android:colorAccent">@color/colorAccent</item>
-        <item name="android:textColor">@color/colorText</item>
-        <item name="android:textColorPrimary">@color/colorPrimaryText</item>
-        <item name="android:textColorSecondary">@color/colorSecondaryText</item>
-        <item name="android:windowBackground">@color/grey95</item>
-        <item name="android:statusBarColor">@color/colorPrimaryDark</item>
+    <!--<style name="AppTheme" parent="Theme.AppCompat.Light.NoActionBar">-->
```
Comment (Astalaseven) [inline app/src/main/res/values/styles.xml]: Remove unused code before merge

**Review by Astalaseven — CHANGES_REQUESTED** [review]

File: app/src/main/assets/github_.css (hunk 1 above)
Comment (fknop) [reply app/src/main/assets/github_.css]: I removed the files for now.

File: app/src/main/java/gitlin/kothub/accounts/LoginActivity.kt (hunk 2 above)
Comment (fknop) [reply app/src/main/java/gitlin/kothub/accounts/LoginActivity.kt:102]: It's from Android API.

File: app/src/main/java/gitlin/kothub/github/api/dsl/Edges.kt (hunk 3 above)
Comment (fknop) [reply app/src/main/java/gitlin/kothub/github/api/dsl/Edges.kt]: Done

File: app/src/main/java/gitlin/kothub/github/api/dsl/Issue.kt (hunk 4 above)
Comment (fknop) [reply app/src/main/java/gitlin/kothub/github/api/dsl/Issue.kt]: Done

File: app/src/main/java/gitlin/kothub/github/api/dsl/User.kt (hunk 5 above)
Comment (fknop) [reply app/src/main/java/gitlin/kothub/github/api/dsl/User.kt]: Done

File: app/src/main/java/gitlin/kothub/ui/drawer/Drawer.kt (hunk 7 above)
Comment (fknop) [reply app/src/main/java/gitlin/kothub/ui/drawer/Drawer.kt]: Done

File: app/src/main/java/gitlin/kothub/ui/repo/RepositoryActivity.kt (hunk 8 above)
Comment (fknop) [reply app/src/main/java/gitlin/kothub/ui/repo/RepositoryActivity.kt]: Done

File: app/src/main/java/gitlin/kothub/ui/repo/RepositoryActivity.kt (hunk 10 above)
Comment (fknop) [reply app/src/main/java/gitlin/kothub/ui/repo/RepositoryActivity.kt]: Done

File: app/src/main/java/gitlin/kothub/ui/settings/SettingsActivity.kt (hunk 11 above)
Comment (fknop) [reply app/src/main/java/gitlin/kothub/ui/settings/SettingsActivity.kt]: Done

File: app/src/main/java/gitlin/kothub/utilities/Fragment.kt (hunk 12 above)
Comment (fknop) [reply app/src/main/java/gitlin/kothub/utilities/Fragment.kt]: Removed the method as it wasn't used.

File: app/src/main/res/values/styles.xml (hunk 15 above)
Comment (fknop) [reply app/src/main/res/values/styles.xml]: Done

File: app/src/main/java/gitlin/kothub/ui/drawer/Drawer.kt [hunk 16]
```diff
@@ -26,25 +19,34 @@ import com.mikepenz.materialdrawer.util.AbstractDrawerImageLoader
...
+import gitlin.kothub.ui.NotificationActivity
+import gitlin.kothub.ui.issues.IssuesActivity
+import gitlin.kothub.ui.pulls.PullRequestsActivity
+import gitlin.kothub.ui.settings.SettingsActivity
 import io.reactivex.rxkotlin.addTo
 import org.jetbrains.anko.*
-import java.text.SimpleDateFormat
 
-inline fun BadgeStyle.whiteText () = this.withTextColor(Color.WHITE)!!
+fun BadgeStyle.whiteText () = this.withTextColor(android.graphics.Color.WHITE)!!
```
Comment (Astalaseven) [inline app/src/main/java/gitlin/kothub/ui/drawer/Drawer.kt]: AS refactor tool for imports seems to have made some undesirable changes

File: app/src/main/java/gitlin/kothub/ui/repo/RepositoryActivity.kt [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+
+class RepositoryViewModel(): ViewModel() {
+
+    val repo = MutableLiveData<RepositorySummary>()
+    val readme = MutableLiveData<Spanned>()
+
+    lateinit var owner: String
+    lateinit var name: String
+
+    val nameWithOwner: String get() = "$owner/$name"
```
Comment (Astalaseven) [inline app/src/main/java/gitlin/kothub/ui/repo/RepositoryActivity.kt:36]: Property is not more used

File: app/src/main/res/layout/repository_header.xml [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+            tools:layout_editor_absoluteX="164dp"
+            tools:layout_editor_absoluteY="8dp" />
+
+        <TextView
+            android:id="@+id/watchers"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_gravity="center_horizontal"
+            android:layout_marginTop="5dp"
+            android:textSize="16dip"
```
Comment (Astalaseven) [inline app/src/main/res/layout/repository_header.xml]: Use `sp` for textSize

File: app/src/main/res/layout/repository_header.xml [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+            tools:layout_editor_absoluteY="8dp" />
+
+
+        <TextView
+            android:id="@+id/stargazers"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_gravity="center_horizontal"
+            android:layout_marginTop="5dp"
+            android:textSize="16dip"
```
Comment (Astalaseven) [inline app/src/main/res/layout/repository_header.xml]: Use `sp` for textSize

File: app/src/main/res/layout/repository_header.xml [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+            tools:layout_editor_absoluteX="292dp" />
+
+        <TextView
+            android:id="@+id/forks"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_gravity="center_horizontal"
+            android:layout_marginBottom="8dp"
+            android:layout_marginTop="5dp"
+            android:textSize="16dip"
```
Comment (Astalaseven) [inline app/src/main/res/layout/repository_header.xml]: Use `sp` for textSize

File: app/src/main/java/gitlin/kothub/LauncherActivity.kt [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    override fun onStart() {
+        super.onStart()
+
+        val authTokenType = getString(R.string.accountType)
+        val am = AccountManager.get(this)
+        val accounts = am.getAccountsByType(authTokenType)
+
+        if (accounts.size > 0) {
```
Comment (Astalaseven) [inline app/src/main/java/gitlin/kothub/LauncherActivity.kt]: `accounts.isNotEmpty()` would be more kotlin idiomatic

File: app/src/main/java/gitlin/kothub/accounts/AccountUtils.kt [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+const val USER_LOGIN = "USER_LOGIN"
+const val USER_NAME = "USER_NAME"
+
+
+fun getAccount (context: Context): Account? {
+    val am = AccountManager.get(context)
+    val authTokenType = context.getString(R.string.accountType)
+    val accounts = am.getAccountsByType(authTokenType)
+
+    if (accounts.size == 0) {
```
Comment (Astalaseven) [inline app/src/main/java/gitlin/kothub/accounts/AccountUtils.kt]: `account.isEmpty()` would be more kotlin idiomatic

**Review by Astalaseven — CHANGES_REQUESTED** [review]

File: app/src/main/java/gitlin/kothub/ui/repo/RepositoryActivity.kt (hunk 17 above)
Comment (fknop) [reply app/src/main/java/gitlin/kothub/ui/repo/RepositoryActivity.kt:36]: I'll let it there for now. It may be useful.

**Review by Astalaseven — APPROVED** [review]

**Pull request merged**